{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with open('/Users/alex/Desktop/files.txt', 'r') as f: \n",
    "    files = f.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "files = [file.strip('\\n').replace('/pool/data/globus/', '') for file in files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for file in files: \n",
    "    \n",
    "    fileparts = file.split('/')\n",
    "    if fileparts[0].split('_')[1] != 'genomics': print(file)\n",
    "    if fileparts[1].split('_')[0] != 'CGND': print(file)\n",
    "        \n",
    "    if fileparts[2][:7] == 'Sample_':\n",
    "        \n",
    "        if not ('.fastq.gz' in fileparts[3] or \n",
    "                 '.bam' in fileparts[3] or \n",
    "                 '.bai' in fileparts[3] or \n",
    "                 '.cram' in fileparts[3] or \n",
    "                 '.crai' in fileparts[3] or \n",
    "                 '.haplotypeCalls.er.raw.vcf.gz' in fileparts[3] or \n",
    "                 '.haplotypeCalls.er.raw.vcf.gz.tbi' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.vcf.gz' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.vcf.gz.tbi' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.vcf.gz' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.vcf.gz.tbi' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.clinical.txt' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.coding.txt' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.coding_rare.txt' in fileparts[3] or \n",
    "                 '.recalibrated.haplotypeCalls.annotated.txt' in fileparts[3] or                 \n",
    "                 '.genomestrip.del.bed' in fileparts[3] or\n",
    "                 '.eh.vcf' in fileparts[3]): \n",
    "            print(file)\n",
    "    \n",
    "    elif fileparts[2] == 'StructuralVariants': continue\n",
    "\n",
    "    elif fileparts[2] == 'ExpansionHunter': continue\n",
    "        \n",
    "    else: print(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dd = defaultdict(set)\n",
    "\n",
    "for file in files: \n",
    "    fileparts = file.split('/')\n",
    "    project = fileparts[0]\n",
    "    quote = fileparts[1]\n",
    "    sample = fileparts[2]\n",
    "\n",
    "    dd[project+'/'+quote+'/'+sample].add(fileparts[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = dict()\n",
    "for sample, files in dict(dd).items():\n",
    "    \n",
    "    d[sample] = {\n",
    "        'fastq': len([file for file in files if '.fastq.gz' in file]),\n",
    "        'bam': len([file for file in files if '.final.bam' in file or '.final.bai' in file]),\n",
    "        'cram': len([file for file in files if '.final.cram' in file or '.final.crai' in file]),\n",
    "        'gVCF': len([file for file in files if '.raw.vcf' in file]),\n",
    "        'VCF': len([file for file in files if 'haplotypeCalls.vcf.gz' in file]),\n",
    "        'annotation': len([file for file in files if '.annotated.' in file]),\n",
    "        'SVariants': len([file for file in files if '.genomestrip.del.bed' in file]),\n",
    "        'ExpansionHunter VCF': len([file for file in files if '.eh.vcf' in file])\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq</th>\n",
       "      <th>bam</th>\n",
       "      <th>cram</th>\n",
       "      <th>gVCF</th>\n",
       "      <th>VCF</th>\n",
       "      <th>annotation</th>\n",
       "      <th>SVariants</th>\n",
       "      <th>ExpansionHunter VCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aals_genomics/CGND_12001/Sample_CGND-HDA-00187</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aals_genomics/CGND_12001/Sample_CGND-HDA-00202</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aals_genomics/CGND_12001/Sample_CGND-HDA-00203</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aals_genomics/CGND_12001/Sample_CGND-HDA-00204</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>aals_genomics/CGND_12001/Sample_CGND-HDA-00205</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                fastq  bam  cram  gVCF  VCF  \\\n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00187     16    0     2     2    2   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00202     16    0     2     2    2   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00203     16    0     2     2    2   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00204     16    0     2     2    2   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00205     16    0     2     2    2   \n",
       "\n",
       "                                                annotation  SVariants  \\\n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00187           6          1   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00202           6          1   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00203           6          1   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00204           6          1   \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00205           6          1   \n",
       "\n",
       "                                                ExpansionHunter VCF  \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00187                    0  \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00202                    0  \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00203                    0  \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00204                    0  \n",
       "aals_genomics/CGND_12001/Sample_CGND-HDA-00205                    0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(d).reindex(['fastq','bam','cram','gVCF','VCF','annotation','SVariants', 'ExpansionHunter VCF']).T\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>fastq</th>\n",
       "      <th>bam</th>\n",
       "      <th>cram</th>\n",
       "      <th>gVCF</th>\n",
       "      <th>VCF</th>\n",
       "      <th>annotation</th>\n",
       "      <th>SVariants</th>\n",
       "      <th>ExpansionHunter VCF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>project</th>\n",
       "      <th>quote</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">aals_genomics</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">CGND_12001</th>\n",
       "      <th>Sample_CGND-HDA-00187</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_CGND-HDA-00202</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_CGND-HDA-00203</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_CGND-HDA-00204</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_CGND-HDA-00205</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                fastq  bam  cram  gVCF  VCF  \\\n",
       "project       quote      sample                                               \n",
       "aals_genomics CGND_12001 Sample_CGND-HDA-00187     16    0     2     2    2   \n",
       "                         Sample_CGND-HDA-00202     16    0     2     2    2   \n",
       "                         Sample_CGND-HDA-00203     16    0     2     2    2   \n",
       "                         Sample_CGND-HDA-00204     16    0     2     2    2   \n",
       "                         Sample_CGND-HDA-00205     16    0     2     2    2   \n",
       "\n",
       "                                                annotation  SVariants  \\\n",
       "project       quote      sample                                         \n",
       "aals_genomics CGND_12001 Sample_CGND-HDA-00187           6          1   \n",
       "                         Sample_CGND-HDA-00202           6          1   \n",
       "                         Sample_CGND-HDA-00203           6          1   \n",
       "                         Sample_CGND-HDA-00204           6          1   \n",
       "                         Sample_CGND-HDA-00205           6          1   \n",
       "\n",
       "                                                ExpansionHunter VCF  \n",
       "project       quote      sample                                      \n",
       "aals_genomics CGND_12001 Sample_CGND-HDA-00187                    0  \n",
       "                         Sample_CGND-HDA-00202                    0  \n",
       "                         Sample_CGND-HDA-00203                    0  \n",
       "                         Sample_CGND-HDA-00204                    0  \n",
       "                         Sample_CGND-HDA-00205                    0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['project'], df['quote'], df['sample'] = df.index.str.split('/').str\n",
    "df = df.set_index(['project','quote','sample'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('/Users/alex/Desktop/df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
